packages/big-design/src/components/Tree/types.ts

```diff
@@ -0,0 +1,42 @@
+export interface TreeProps<T> {
+  iconless?: boolean;
+  nodes: TreeNodeProps<T>[];
+  selectable?: 'radio' | 'multi';
+  onExpand?(node: TreeNodeProps<T>): TreeNodeRef<T> | void;
+  onCollapse?(node: TreeNodeProps<T>): TreeNodeRef<T> | void;
+  // Based on the selectable prop, it can either
+  // return and array of values, or just one.
+  onSelect?(values: T[] | T): void;
+}
+
+export type TreeNodeId = string | number;
+
+export interface TreeNodeProps<T> {
+  children?: TreeNodeProps<T>[];
+  disabled?: boolean;
+  expanded?: boolean;
+  icon?: React.ReactChild | null;
+  id: TreeNodeId;
+  label: string;
+  selected?: boolean;
+  value?: T;
+}
+
+export type TreeNodeRef<T> = Pick<TreeNodeProps<T>, 'children'>;
+
+export type NodeMap = {
+  [nodeId in TreeNodeId]: {
+    children: TreeNodeId[];
+    id: TreeNodeId;
+    parent: TreeNodeId;
+  };
+};
+
+export interface TreeState<T> {
+  nodeMap: NodeMap;
+  expandedNodeIds: TreeNodeId[];
+  flattenedNodeIds: TreeNodeId[];
+  focusedNode: TreeNodeId;
+  selectedValues: Array<T>;
+  visibleNodeIds: TreeNodeId[];
```

Comment:
Other than `flattenedNodeIds` I don't think the others need to be arrays. Maybe we can use `Set` here 🤔. This should avoid having to iterate everywhere.

Reply:
I thought about doing that at the beginning. I dropped the idea, but I'll try it out after resolving the above comments.